lib/puppet/provider/cs_stonith/pcs.rb

```diff
@@ -0,0 +1,209 @@
+      :operations               => {},
+      :existing_resource        => :true,
+      :existing_primitive_type  => e.attributes['type'],
+      :existing_operations      => {},
+    }
+
+    if ! e.elements['operations'].nil?
+      e.elements['operations'].each_element do |o|
+        valids = o.attributes.reject do |k,v| k == 'id' end
+        if ! valids['role'].nil?
+          name = valids['name']
+          name << ":"
+          name << valids['role']
+        else
+          name = valids['name']
+        end
+        hash[:operations][name] = {}
+        valids.each do |k,v|
+          hash[:operations][name][k] = v if k != 'name' and k != 'role'
+        end
+        hash[:existing_operations] = hash[:operations].dup
+      end
+    end
+    hash
+  end
+
+  def self.instances
+
+    block_until_ready
+
+    instances = []
+
+    cmd = [ command(:pcs), 'cluster', 'cib' ]
+    raw, status = run_pcs_command(cmd)
+    doc = REXML::Document.new(raw)
+
+    # What does this mean? Is stonith correct?
+    REXML::XPath.each(doc, '//primitive') do |e|
+      instances << new(element_to_hash(e))
+    end
+    instances
+  end
+
+  # Create just adds our resource to the property_hash and flush will take care
+  # of actually doing the work.
+  # The existing_resource is there because pcs does not have a single command that
+  # updates or create a resource, so we flag the resources with that parameter
+  def create
+    @property_hash = {
+      :name              => @resource[:name],
+      :ensure            => :present,
+      :primitive_type    => @resource[:primitive_type],
+      :existing_resource => :false
+    }
+    @property_hash[:device_options] = @resource[:device_options] if ! @resource[:device_options].nil?
+    @property_hash[:operations] = @resource[:operations] if ! @resource[:operations].nil?
+  end
+
+  # Unlike create we actually immediately delete the item.
+  def destroy
+    debug('Removing STONITH fence')
+    Puppet::Provider::Pacemaker::run_pcs_command([command(:pcs), 'stonith', 'delete', @property_hash[:name]])
+    @property_hash.clear
+  end
+
+  # Getters that obtains the device_options and operations defined in our primitive
+  # that have been populated by prefetch or instances (depends on if your using
+  # puppet resource or not).
+  def primitive_type
+    @property_hash[:primitive_type]
+  end
+
+  def device_options
+    @property_hash[:device_options]
+  end
+
+  def operations
+    @property_hash[:operations]
+  end
+
+  # Our setters for device_options and operations.  Setters are used when the
+  # resource already exists so we just update the current value in the
+  # property_hash and doing this marks it to be flushed.
+  def fence_type=(should)
+    @property_hash[:fence_type] = should
+  end
+
+  def device_options=(should)
+    @property_hash[:device_options] = should
+  end
+
+  def operations=(should)
+    @property_hash[:operations] = should
+  end
```

Comment:
I'll change. Don't be confuzzled. I didn't know about mk_resource_methods until you mentioned it.